### PR TITLE
fix(notifications): Fix `get_user_subscriptions_for_groups()`

### PR DESCRIPTION
**Pull request by mgaeta**
When a user has a more specific "OFF" value than one of the "ON" values, we're incorrectly marking the group as "OFF".
```
type        provider    scope   value
---------------------------------------
workflow    slack       user    ON
workflow    email       user    OFF
workflow    email       project OFF
```

### CODE REVIEW

File: src/sentry/notifications/helpers.py [hunk 1]
```diff
@@ -472,3 +473,61 @@ def get_most_specific_notification_setting_value(
         )
         or _get_notification_setting_default(ExternalProviders.EMAIL, type)
     )
+
+
+def are_there_any_settings(
```
Comment (AniketDas-Tekky) [inline src/sentry/notifications/helpers.py:478]: If this is supposed to be a library function, can you add docstrings?

File: src/sentry/notifications/helpers.py [hunk 2]
```diff
@@ -472,3 +473,61 @@ def get_most_specific_notification_setting_value(
...
+    )
+    return any(
+        [
+            value != NotificationSettingOptionValues.NEVER
+            for value in highest_value_by_provider.values()
+        ]
+    )
+
+
+def get_highest_values_by_provider(
```
Comment (AniketDas-Tekky) [inline src/sentry/notifications/helpers.py:501]: please add docstrings

File: src/sentry/notifications/helpers.py [hunk 3]
```diff
@@ -472,3 +473,61 @@ def get_most_specific_notification_setting_value(
...
+    notification_settings_by_scope: Mapping[
+        NotificationScopeType,
+        Mapping[int, Mapping[ExternalProviders, NotificationSettingOptionValues]],
+    ],
+    recipient: Team | User,
+    parent_id: int,
+    type: NotificationSettingTypes,
+) -> Mapping[ExternalProviders, NotificationSettingOptionValues]:
+    highest_value_by_provider = {
+        provider: _get_notification_setting_default(provider, type)
```
Comment (scefali) [inline src/sentry/notifications/helpers.py:511]: @mgaeta I'm not quite understanding why the highest value enum takes priority over another enum.

File: src/sentry/notifications/helpers.py [hunk 4]
```diff
@@ -472,3 +473,61 @@ def get_most_specific_notification_setting_value(
...
+        for provider in notification_providers()
+    }
+
+    parent_notification_settings_by_provider = notification_settings_by_scope.get(
+        get_scope_type(type), {}
+    ).get(parent_id, {})
+    for provider, value in parent_notification_settings_by_provider.items():
```
Comment (scefali) [inline src/sentry/notifications/helpers.py:518]: @mgaeta can you add comments on what's going on inside this function?

File: src/sentry/notifications/helpers.py (hunk 1 above)
Comment (ceorourke) [reply src/sentry/notifications/helpers.py:478]: What do you think of calling this `has_settings`?

File: src/sentry/notifications/helpers.py [hunk 5]
```diff
@@ -461,14 +501,65 @@ def get_most_specific_notification_setting_value(
...
+
+
+def merge_notification_settings_up(
+    *settings_mappings: Mapping[ExternalProviders, NotificationSettingOptionValues],
+) -> Mapping[ExternalProviders, NotificationSettingOptionValues]:
+    """
+    Given a list of notification settings by provider ordered by increasing
+    specificity, get the most specific value by provider.
+    """
+    value_by_provider: MutableMapping[ExternalProviders, NotificationSettingOptionValues] = {}
```
Comment (ceorourke) [inline src/sentry/notifications/helpers.py:541]: I'm having a hard time understanding what's happening here, can you give an example of an input and the changed output? Where is the ordering by increasing specificity happening?

**Comment by ceorourke** [comment]
Based on the Slack thread I think the db representation in the description does not match the problem, he said the projects were enabled so it's more like this:

```
type        provider    scope   value
---------------------------------------
workflow    slack       user    ON
workflow    email       user    OFF
workflow    email       project ON
```

(this may be a bit of stream of consciousness as I try to understand this PR, but I wanted to get my thoughts down)

File: src/sentry/notifications/helpers.py [hunk 6]
```diff
@@ -446,6 +453,39 @@ def get_highest_notification_setting_value(
     return max(notification_settings_by_provider.values(), key=lambda v: v.value)
 
 
+def get_value_for_parent(
```
Comment (ceorourke) [inline src/sentry/notifications/helpers.py:456]: Note to myself as I review, this and `get_value_for_actor` are not related to the bug fix, this is just pulling them out into their own functions.

File: src/sentry/notifications/helpers.py (hunk 5 above)
Comment (mgaeta) [reply src/sentry/notifications/helpers.py:541]:
```py
a = {'x': 1, 'z': 1}
b = {'x': 2, 'y': 2}

 # {'x': 2, 'y': 2, 'z': 1}
merge_notification_settings_up(a, b)

 # {'x': 1, 'y': 2, 'z': 1}
merge_notification_settings_up(b, a)
```

**Comment by mgaeta** [comment]
>Based on the Slack thread I think the db representation in the description does not match the problem,

So the DB state in the PR description is correct, but the UI is correctly interpreting it as "delivery method is slack" and project-specific settings are "Default (On)".

**Comment by ceorourke** [comment]
> > Based on the Slack thread I think the db representation in the description does not match the problem,
> 
> So the DB state in the PR description is correct, but the UI is correctly interpreting it as "delivery method is slack" and project-specific settings are "Default (On)".

Do you mean incorrectly interpreting or correctly?

File: src/sentry/notifications/helpers.py (hunk 5 above)
Comment (ceorourke) [reply src/sentry/notifications/helpers.py:539]: Okay I'm still a little confused, `a` in this case is always whatever default values we've hardcoded, and then `b` is all of the user's settings, is that correct? So as a concrete example I ran [test_get_setting_mapping_from_mapping_project](https://github.com/getsentry/sentry/blob/73294e450a8cd61340b50ec340123c7901ec3286/tests/sentry/notifications/utils/test_get_setting_mapping_from_mapping.py#L118) which I believe has user settings that are the same as the problem this PR is solving and then printed this out:
```
default_settings = {EMAIL: ALWAYS, SLACK: NEVER}
user_settings = {EMAIL: NEVER: 10, SLACK: ALWAYS}, {}, {EMAIL: NEVER}
result = merge_notification_settings_up(default_settings, user_settings)
result: {EMAIL: NEVER, SLACK: ALWAYS}
```

I guess this isn't the most helpful example because there is no change though. So is the default value being passed so that it can be used if there happen to be no settings for a given provider? I also still don't understand where the ordering by increasing specificity is happening.

To sum up: beforehand the project level `{EMAIL: NEVER}` was taking priority over the user level `{SLACK: ALWAYS}`, causing the group subscription (the bell icon) to be disabled, and now we are instead (I need to save this comment and re-read the changes) checking the user level settings?

**Comment by mgaeta** [comment]
> > > Based on the Slack thread I think the db representation in the description does not match the problem,
> > 
> > So the DB state in the PR description is correct, but the UI is correctly interpreting it as "delivery method is slack" and project-specific settings are "Default (On)".
> 
> Do you mean incorrectly interpreting or correctly?
correctly

**Review by ceorourke — APPROVED** [review]
